File: 2022/day-5/day5.py
import re
from operator import methodcaller as method


def parse_input(inp: str):
    stacks_map, instructions = map(method('splitlines'), inp.split('\n\n'))
    stacks_ids = [int(m)for m in re.findall(r'\d+', stacks_map[-1])]
    stacks = {sid: [] for sid in stacks_ids}
    create_regex = re.compile(r'(\s{3}|\[[A-Z]\])\s?')
    for line in reversed(stacks_map[:-1]):
        for stack_id, crate in zip(stacks_ids, create_regex.findall(line)):
            if crate.strip():
                stacks[stack_id].insert(0, crate[1])
    instruction_regex = re.compile(r'move\s(\d+)\sfrom\s(\d+)\sto\s(\d+)')
    return stacks, [tuple(map(int, instruction_regex.match(inst).groups())) for inst in instructions]


def part_one(stacks, instructions):
    for amount, src, dst in instructions:
        for _ in range(amount):
            crate = stacks[src].pop(0)
            stacks[dst].insert(0, crate)
    print(''.join(stacks[sid][0] for sid in sorted(stacks)))


def part_two(stacks, instructions):
    for amount, src, dst in instructions:
        crates = stacks[src][:amount]
        del stacks[src][:amount]
        for crate in reversed(crates):
            stacks[dst].insert(0, crate)
    print(''.join(stacks[sid][0] for sid in sorted(stacks)))


with open("input.txt") as h:
    contents = h.read()
    part_one(*parse_input(contents))
    part_two(*parse_input(contents))

File: 2022/day-5/day5.rs
use regex::Regex;
use std::collections::HashMap;

type Stacks = HashMap<usize, Vec<char>>;
type Instruction = (usize, usize, usize);

fn parse_input(inp: &str) -> (Stacks, Vec<Instruction>) {
    let mut main_sep = inp.split("\n\n");
    let stack_map: Vec<&str> = main_sep.next().unwrap().lines().collect();
    let regex_stack_ids = Regex::new(r"(\d+)").unwrap();
    let stack_ids: Vec<usize> = regex_stack_ids
        .find_iter(stack_map.last().unwrap())
        .map(|m| m.as_str().parse::<usize>().unwrap())
        .collect();
    let mut stacks: Stacks = HashMap::from_iter(stack_ids.iter().map(|sid| (*sid, Vec::new())));
    let create_regex = Regex::new(r"(\s{3}|\[[A-Z]\])\s?").unwrap();
    for &line in stack_map.iter().take(stack_map.len() - 1) {
        for (stack_id, m) in stack_ids.iter().zip(create_regex.find_iter(line)) {
            if !m.as_str().trim().is_empty() {
                stacks
                    .get_mut(stack_id)
                    .expect("Invalid stack Id")
                    .push(*(&m.as_str()[1..2].chars().next().unwrap()));
            }
        }
    }
    let instructions = main_sep.next().unwrap();
    let instructions: Vec<Instruction> = instructions
        .lines()
        .map(|l| {
            let v: Vec<usize> = l
                .split(' ')
                .skip(1)
                .step_by(2)
                .map(|c| c.parse::<usize>().unwrap())
                .collect();
            (v[0], v[1], v[2])
        })
        .collect();
    (stacks, instructions)
}

fn upper_crates(stacks: &Stacks) -> Vec<char> {
    (1..stacks.keys().max().unwrap() + 1)
        .map(|id| stacks.get(&(id as usize)).unwrap().get(0).unwrap())
        .cloned()
        .collect()
}

fn part_one(stacks: &mut Stacks, instructions: Vec<Instruction>) {
    for (amount, src, dest) in instructions {
        for _ in 0..amount {
            let cr = stacks.get_mut(&src).unwrap().remove(0);
            stacks.get_mut(&dest).unwrap().insert(0, cr);
        }
    }
    println!("Part 1: {:?}", upper_crates(&stacks));
}

fn part_two(stacks: &mut Stacks, instructions: Vec<Instruction>) {
    for (amount, src, dst) in instructions {
        let crates: Vec<char> = (0..amount)
            .map(|i| stacks.get_mut(&src).unwrap().remove(0))
            .collect();
        crates
            .iter()
            .rev()
            .for_each(|&c| stacks.get_mut(&dst).unwrap().insert(0, c))
    }
    println!("Part 2: {:?}", upper_crates(&stacks));
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input");
    let (stacks, instructions) = parse_input(&contents);
    part_one(&mut stacks.clone(), instructions.clone());
    part_two(&mut stacks.clone(), instructions);
}
